### PR TITLE
Use replace instead of patch for kubectl edit

### LINKED ISSUE #35016: Kubectl edit should use replace instead of patch
Currently, when using kubectl edit, the end user is presented with the full yaml, makes some changes after looking at it, and then patches the existing yaml.  Just doing a replace and throwing a user facing error on a optimistic lock failure seems like a better approach since this is user facing and the full yaml is present.

### PR DESCRIPTION
**Pull request by mengqiy**
Use replace instead of patch for `kubectl edit`.
Fixes: #35016, #32398
But the root cause of #32398, which is related to strategic patch, is not fixed by this PR.
@pwittrock

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35171)

### CODE REVIEW

**Comment by ymqytw** [comment]
@k8s-bot cvm gke e2e test this

**Comment by ymqytw** [comment]
@k8s-bot gci gke e2e test this

File: pkg/kubectl/cmd/edit.go [hunk 1]
```diff
@@ -396,56 +395,50 @@ func visitToPatch(originalObj runtime.Object, updates *resource.Info, mapper met
 
 		}
 
-		originalSerialization, err := runtime.Encode(encoder, currOriginalObj)
+		ok, err := checkAPiVersionKindName(currOriginalObj, info.Object)
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/edit.go]: Or try a more meaningful name, something like 

```
valid, err := validateImmutableAPIFields(currOriginalObj, info.Object)
```

File: pkg/kubectl/cmd/edit.go [hunk 2]
```diff
@@ -19,11 +19,11 @@ package cmd
 import (
 	"bufio"
 	"bytes"
+	goerrors "errors"
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/edit.go]: Just leave this as `"errors"`.

File: pkg/kubectl/cmd/edit.go [hunk 3]
```diff
@@ -396,56 +395,50 @@ func visitToPatch(originalObj runtime.Object, updates *resource.Info, mapper met
...
-			strategicpatch.RequireKeyUnchanged("kind"), strategicpatch.RequireMetadataKeyUnchanged("name")}
-		patch, err := strategicpatch.CreateTwoWayMergePatch(originalJS, editedJS, currOriginalObj, preconditions...)
-		if err != nil {
-			glog.V(4).Infof("Unable to calculate diff, no merge is possible: %v", err)
-			if strategicpatch.IsPreconditionFailed(err) {
-				return preservedFile(nil, file, errOut)
-			}
-			return err
+		if !ok {
+			return goerrors.New("At least one of apiVersion, kind and name was changed")
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/edit.go]: Not very clear, try something like "The following fields must not be changed: apiVersion, kind, and metadata.name".

**Review by fabianofranz — CHANGES_REQUESTED** [review]

**Comment by zefciu** [comment]
According to the discussion in https://github.com/kubernetes/kubernetes/issues/32398 this is not the preferred way to fix this bug. We should rather fix the PATCH verb.

**Comment by ymqytw** [comment]
We are actively fixing the strategic patch that used by PATCH.

**Comment by pwittrock** [comment]
@zefciu Why should we be using PATCH for kubectl edit since the whole config is read into the editor and visible?  Since this is consumed by an end user, wouldn't we prefer getting an optimistic lock exception if something changes anyway?

What was the original rationale for using patch for kubectl edit?

**Comment by ymqytw** [comment]
@zefciu I created a PR #35647 to fix the root cause of #32398.

**Comment by ymqytw** [comment]
@fabianofranz Updated. PTAL

**Comment by fabianofranz** [comment]
I'm not completely confident about this. 

Replace seems to be a better fit than patch for `edit` since you have the entire object marshalled in the editor in front of you, and I understand that when you hit `:wq` you are deliberately saying "take it, this is how my object should look like", not "take this piece I just changed, not the whole of what's in the screen". If the entire object could not be replaced because it changed in the meantime, you get an error (optimistic lock failure).

On the other hand, @smarterclayton has a point in https://github.com/kubernetes/kubernetes/issues/32398#issuecomment-250620305 when saying there are objects that change rapidly and multiple times which would in some cases even prevent you from using `edit` if we do a full replace. Which may be the correct thing to do, just annoying.

@ymqytw does https://github.com/kubernetes/kubernetes/pull/35647 completely fix editing when you remove elements from a list (https://github.com/kubernetes/kubernetes/issues/32398)?

**Comment by ymqytw** [comment]
> there are objects that change rapidly and multiple times which would in some cases even prevent you from using edit if we do a full replace. Which may be the correct thing to do, just annoying.

This is safe, but sometimes could be annoying.

> does #35647 completely fix editing when you remove elements from a list (#32398)?

I still working on #35647. But I believe it will completely fix the commands using strategicMergePatch, i.e. apply, edit and patch.

**Comment by smarterclayton** [comment]
We explicitly put patch in because the experience of edit without it was
unusable on anything that was under active controller action.

Replace should never really be used in most clients.  I'm ok with making it
be an option to edit, but it's not a great story in general.

**Comment by ymqytw** [comment]
Closed this PR.
Ref: #35016

**Pull request closed without merging**